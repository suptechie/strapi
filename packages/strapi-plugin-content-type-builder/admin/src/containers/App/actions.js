/*
 *
 * App actions
 *
 */
import { pick, set, camelCase } from 'lodash';
import {
  ADD_ATTRIBUTE_TO_EXISITING_CONTENT_TYPE,
  ADD_ATTRIBUTE_TO_TEMP_CONTENT_TYPE,
  CANCEL_NEW_CONTENT_TYPE,
  CLEAR_TEMPORARY_ATTRIBUTE,
  CREATE_TEMP_CONTENT_TYPE,
  DELETE_MODEL,
  DELETE_MODEL_ATTRIBUTE,
  DELETE_MODEL_SUCCEEDED,
  DELETE_TEMPORARY_MODEL,
  GET_DATA,
  GET_DATA_SUCCEEDED,
  ON_CHANGE_NEW_CONTENT_TYPE_MAIN_INFOS,
  ON_CREATE_ATTRIBUTE,
  RESET_NEW_CONTENT_TYPE_MAIN_INFOS,
  RESET_EDIT_EXISTING_CONTENT_TYPE,
  RESET_EXISTING_CONTENT_TYPE_MAIN_INFOS,
  RESET_EDIT_TEMP_CONTENT_TYPE,
  RESET_PROPS,
  SAVE_EDITED_ATTRIBUTE,
  SET_TEMPORARY_ATTRIBUTE,
  SUBMIT_TEMP_CONTENT_TYPE,
  SUBMIT_TEMP_CONTENT_TYPE_SUCCEEDED,
  UPDATE_TEMP_CONTENT_TYPE,
  ON_CHANGE_EXISTING_CONTENT_TYPE_MAIN_INFOS,
} from './constants';

export function addAttributeToExistingContentType(contentTypeName, attributeType) {
  return {
    type: ADD_ATTRIBUTE_TO_EXISITING_CONTENT_TYPE,
    attributeType,
    contentTypeName,
  };
}

export function addAttributeToTempContentType(attributeType) {
  return {
    type: ADD_ATTRIBUTE_TO_TEMP_CONTENT_TYPE,
    attributeType,
  };
}

export function cancelNewContentType() {
  return {
    type: CANCEL_NEW_CONTENT_TYPE,
  };
}

export function clearTemporaryAttribute() {
  return {
    type: CLEAR_TEMPORARY_ATTRIBUTE,
  };
}

export function createTempContentType() {
  return {
    type: CREATE_TEMP_CONTENT_TYPE,
  };
}

export function deleteModel(modelName) {
  return {
    type: DELETE_MODEL,
    modelName,
  };
}

export function deleteModelAttribute(keys) {
  return {
    type: DELETE_MODEL_ATTRIBUTE,
    keys,
  };
}

export function deleteModelSucceeded(modelName) {
  return {
    type: DELETE_MODEL_SUCCEEDED,
    modelName,
  };
}

export function deleteTemporaryModel() {
  return {
    type: DELETE_TEMPORARY_MODEL,
  };
}

export function getData() {
  return {
    type: GET_DATA,
  };
}

export function getDataSucceeded({ allModels, models }, connections) {
  const initialData = allModels.reduce((acc, current) => {
    acc[current.name] = pick(current, ['name', 'collectionName', 'connection', 'description', 'mainField']);
    const attributes = buildModelAttributes(current.attributes);
    set(acc, [current.name, 'attributes'], attributes);

    return acc;
  }, {});

  return {
    type: GET_DATA_SUCCEEDED,
    initialData,
    models,
    connections,
  };
}

export function onChangeExistingContentTypeMainInfos({ target }) {
  const value = target.name === 'name' ? camelCase(target.value.trim()).toLowerCase() : target.value.trim();

  return {
    type: ON_CHANGE_EXISTING_CONTENT_TYPE_MAIN_INFOS,
    keys: target.name.split('.'),
    value,
  };
}

export function onChangeNewContentTypeMainInfos({ target }) {
  const value = target.name === 'name' ? camelCase(target.value.trim()).toLowerCase() : target.value.trim();

  return {
    type: ON_CHANGE_NEW_CONTENT_TYPE_MAIN_INFOS,
    keys: target.name.split('.'),
    value,
  };
}

export function onCreateAttribute({ target }) {
  return {
    type: ON_CREATE_ATTRIBUTE,
    keys: target.name.split('.'),
    value: target.value,
  };
}

export function saveEditedAttribute(attributeName, isModelTemporary, modelName) {
  return {
    type: SAVE_EDITED_ATTRIBUTE,
    attributeName,
    isModelTemporary,
    modelName,
  };
}

export function setTemporaryAttribute(attributeName, isModelTemporary, modelName) {
  return {
    type: SET_TEMPORARY_ATTRIBUTE,
    attributeName,
    isModelTemporary,
    modelName,
  };
}

export function resetNewContentTypeMainInfos() {
  return {
    type: RESET_NEW_CONTENT_TYPE_MAIN_INFOS,
  };
}

export function resetEditExistingContentType(contentTypeName) {
  return {
    type: RESET_EDIT_EXISTING_CONTENT_TYPE,
    contentTypeName,
  };
}

export function resetExistingContentTypeMainInfos(contentTypeName) {
  return {
    type: RESET_EXISTING_CONTENT_TYPE_MAIN_INFOS,
    contentTypeName,
  };
}

export function resetEditTempContentType() {
  return {
    type: RESET_EDIT_TEMP_CONTENT_TYPE,
  };
}

export function resetProps() {
  return {
    type: RESET_PROPS,
  };
}

export function submitTempContentType() {
  return {
    type: SUBMIT_TEMP_CONTENT_TYPE,
  };
}

export function submitTempContentTypeSucceeded() {
  return {
    type: SUBMIT_TEMP_CONTENT_TYPE_SUCCEEDED,
  };
}

export function updateTempContentType() {
  return {
    type: UPDATE_TEMP_CONTENT_TYPE,
  };
}

// utils
export const buildModelAttributes = attributes => {
  const formattedAttributes = attributes.reduce((acc, current) => {
    acc[current.name] = current.params;

    return acc;
  }, {});

  return formattedAttributes;
};
