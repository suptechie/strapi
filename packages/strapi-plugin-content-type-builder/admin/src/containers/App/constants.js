/*
 *
 * App constants
 *
 */

export const DELETE_CONTENT_TYPE = 'ContentTypeBuilder/App/DELETE_CONTENT_TYPE';
export const MODELS_FETCH = 'ContentTypeBuilder/App/MODELS_FETCH';
export const MODELS_FETCH_SUCCEEDED = 'ContentTypeBuilder/App/MODELS_FETCH_SUCCEEDED';
export const STORE_TEMPORARY_MENU = 'ContentTypeBuilder/App/STORE_TEMPORARY_MENU';
