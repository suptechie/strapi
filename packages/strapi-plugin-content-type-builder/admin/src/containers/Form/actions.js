/*
 *
 * Form actions
 *
 */

 /* eslint-disable new-cap */

import { map , forEach, replace } from 'lodash';
import { Map, List } from 'immutable';
import { getValidationsFromForm } from '../../utils/formValidations';
import { storeData } from '../../utils/storeData';

import {
  CHANGE_INPUT,
  CHANGE_INPUT_ATTRIBUTE,
  CONNECTIONS_FETCH,
  CONNECTIONS_FETCH_SUCCEEDED,
  CONTENT_TYPE_ACTION_SUCCEEDED,
  CONTENT_TYPE_CREATE,
  CONTENT_TYPE_EDIT,
  CONTENT_TYPE_FETCH,
  CONTENT_TYPE_FETCH_SUCCEEDED,
  REMOVE_CONTENT_TYPE_REQUIRED_ERROR,
  RESET_DID_FETCH_MODEL_PROP,
  RESET_FORM_ERRORS,
  RESET_IS_FORM_SET,
  SET_ATTRIBUTE_FORM,
  SET_ATTRIBUTE_FORM_EDIT,
  SET_BUTTON_LOADING,
  SET_FORM,
  SET_FORM_ERRORS,
  UNSET_BUTTON_LOADING,
} from './constants';

import forms from './forms.json';

export function changeInput(key, value, isEditing) {
  const objectToModify = isEditing ? 'modifiedDataEdit' : 'modifiedData';
  return {
    type: CHANGE_INPUT,
    key,
    value,
    objectToModify,
  };
}

export function changeInputAttribute(key, value) {
  const keys = key.split('.');
  const firstKey = keys[0];
  const secondKey = keys[1];
  return {
    type: CHANGE_INPUT_ATTRIBUTE,
    firstKey,
    secondKey,
    value,
  };
}

export function connectionsFetch() {
  return {
    type: CONNECTIONS_FETCH,
  };
}

export function connectionsFetchSucceeded(data) {
  const connections = map(data.connections, (connection) => ({ name: connection, value: connection }));
  return {
    type: CONNECTIONS_FETCH_SUCCEEDED,
    connections,
  };
}

export function contentTypeActionSucceeded() {
  return {
    type: CONTENT_TYPE_ACTION_SUCCEEDED,
  };
}

export function contentTypeCreate(newModel) {
  storeData.setContentType(newModel);

  return {
    type: CONTENT_TYPE_CREATE,
  };
}

export function contentTypeFetch(contentTypeName) {
  return {
    type: CONTENT_TYPE_FETCH,
    contentTypeName,
  };
}

export function contentTypeFetchSucceeded(contentType) {
  const dataArray = [['attributes', List(contentType.model.attributes)]];

  forEach(contentType.model, (value, key) => {
    if (key !== 'attributes') {
      dataArray.push([key, value]);
    }
  });

  const data = Map(dataArray);
  return {
    type: CONTENT_TYPE_FETCH_SUCCEEDED,
    data,
  };
}

export function contentTypeEdit() {
  return {
    type: CONTENT_TYPE_EDIT,
  };
}

export function removeContentTypeRequiredError() {
  return {
    type: REMOVE_CONTENT_TYPE_REQUIRED_ERROR,
  };
}

export function resetDidFetchModelProp() {
  return {
    type: RESET_DID_FETCH_MODEL_PROP,
  };
}

export function resetFormErrors() {
  return {
    type: RESET_FORM_ERRORS,
  };
}

export function resetIsFormSet() {
  return {
    type: RESET_IS_FORM_SET,
  };
}

export function setAttributeForm(hash) {
  const data = setAttributeFormData(hash);
  const formValidations = getValidationsFromForm(data.form, []);
  return {
    type: SET_ATTRIBUTE_FORM,
    form: data.form,
    attribute: data.attribute,
    formValidations,
  }
}

export function setAttributeFormEdit(hash, contentType) {
  const form = setAttributeFormData(hash).form;
  const contentTypeAttribute = contentType.attributes[hash.split('::')[3]];
  const formValidations = getValidationsFromForm(form, []);
  const attribute = Map({
    name: contentTypeAttribute.name,
    params: Map(contentTypeAttribute.params),
  });

  return {
    type: SET_ATTRIBUTE_FORM_EDIT,
    form,
    attribute,
    formValidations,
  }
}

export function setButtonLoading() {
  return {
    type: SET_BUTTON_LOADING,
  };
}

export function setForm(hash) {
  const form = forms[hash.split('::')[1]][hash.split('::')[2]];
  const data = getDataFromForm(forms[hash.split('::')[1]]);
  const formValidations = getValidationsFromForm(forms[hash.split('::')[1]], [])

  return {
    type: SET_FORM,
    form,
    data,
    formValidations,
  };
}


export function setFormErrors(formErrors) {
  return {
    type: SET_FORM_ERRORS,
    formErrors,
  };
}

export function unsetButtonLoading() {
  return {
    type: UNSET_BUTTON_LOADING,
  };
}





/**
*
* @param  {object} form
* @return {object} data : An object { target: value }
*/

function getDataFromForm(form) {
  const dataArray = [['attributes', List()]];

  forEach(form, (formSection) => {
    map(formSection.items, (item) => dataArray.push([item.target, item.value]));
  });

  const data = Map(dataArray);

  return data;
}

function setAttributeFormData(hash) {
  const hashArray = hash.split('::');
  const formType = replace(hashArray[1], 'attribute', '');
  const settingsType = hashArray[2];
  const form = forms.attribute[formType][settingsType];

  const attribute = Map({
    name: '',
    params: Map({
      type: formType,
      required: false,
      maxLength: false,
      minLength: false,
    }),
  });

  return {
    form,
    attribute,
  }
}
