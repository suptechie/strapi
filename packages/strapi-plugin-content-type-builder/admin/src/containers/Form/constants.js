/*
 *
 * Form constants
 *
 */

export const CHANGE_INPUT = 'ContentTypeBuilder/Form/CHANGE_INPUT';
export const CHANGE_INPUT_ATTRIBUTE = 'ContentTypeBuilder/Form/CHANGE_INPUT_ATTRIBUTE';
export const CONNECTIONS_FETCH = 'ContentTypeBuilder/Form/CONNECTIONS_FETCH';
export const CONNECTIONS_FETCH_SUCCEEDED = 'ContentTypeBuilder/Form/CONNECTIONS_FETCH_SUCCEEDED';
export const CONTENT_TYPE_ACTION_SUCCEEDED = 'ContentTypeBuilder/Form/CONTENT_TYPE_ACTION_SUCCEEDED';
export const CONTENT_TYPE_CREATE = 'ContentTypeBuilder/Form/CONTENT_TYPE_CREATE';
export const CONTENT_TYPE_EDIT = 'ContentTypeBuilder/Form/CONTENT_TYPE_EDIT';
export const CONTENT_TYPE_FETCH = 'ContentTypeBuilder/Form/CONTENT_TYPE_FETCH';
export const CONTENT_TYPE_FETCH_SUCCEEDED = 'ContentTypeBuilder/Form/CONTENT_TYPE_FETCH_SUCCEEDED';
export const RESET_DID_FETCH_MODEL_PROP = 'ContentTypeBuilder/Form/RESET_DID_FETCH_MODEL_PROP';
export const SET_ATTRIBUTE_FORM = 'ContentTypeBuilder/Form/SET_ATTRIBUTE_FORM';
export const SET_FORM = 'ContentTypeBuilder/Form/SET_FORM';
