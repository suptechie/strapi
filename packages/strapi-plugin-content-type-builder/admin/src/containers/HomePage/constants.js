/*
 *
 * HomePage constants
 *
 */

export const DEFAULT_ACTION = 'src/HomePage/DEFAULT_ACTION';
