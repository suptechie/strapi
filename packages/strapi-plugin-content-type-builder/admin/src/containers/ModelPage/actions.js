/*
 *
 * ModelPage actions
 *
 */
import { forEach, get, includes, map, set } from 'lodash';
import { storeData } from '../../utils/storeData';

import {
  ADD_ATTRIBUTE_TO_CONTENT_TYPE,
  EDIT_CONTENT_TYPE_ATTRIBUTE,
  CANCEL_CHANGES,
  DEFAULT_ACTION,
  DELETE_ATTRIBUTE,
  MODEL_FETCH,
  MODEL_FETCH_SUCCEEDED,
  POST_CONTENT_TYPE_SUCCEEDED,
  SET_BUTTON_LOADER,
  SUBMIT,
  RESET_SHOW_BUTTONS_PROPS,
  UNSET_BUTTON_LOADER,
  UPDATE_CONTENT_TYPE,
} from './constants';

export function addAttributeToContentType(newAttribute) {
  return {
    type: ADD_ATTRIBUTE_TO_CONTENT_TYPE,
    newAttribute,
  };
}

export function editContentTypeAttribute(modifiedAttribute, attributePosition) {
  return {
    type: EDIT_CONTENT_TYPE_ATTRIBUTE,
    modifiedAttribute,
    attributePosition,
  };
}

export function cancelChanges() {
  return {
    type: CANCEL_CHANGES,
  };
}

export function deleteAttribute(position, modelName) {
  const temporaryContentType = storeData.getContentType();
  let sendRequest = true;
  if (get(temporaryContentType, 'name') === modelName) {
    sendRequest = false;
    temporaryContentType.attributes.splice(position, 1);
    const updatedContentType = temporaryContentType;
    storeData.setContentType(updatedContentType);
  }

  return {
    type: DELETE_ATTRIBUTE,
    position,
    sendRequest,
    modelName,
  };
}

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}

export function modelFetch(modelName) {
  return {
    type: MODEL_FETCH,
    modelName,
  };
}

export function modelFetchSucceeded(data) {
  const model = data;
  const defaultKeys = ['required', 'unique', 'type'];

  forEach(model.model.attributes, (attribute, index) => {
    map(attribute.params, (value, key) => {
      if (!includes(defaultKeys, key) && value) {
        set(model.model.attributes[index].params, `${key}Value`, value);
        set(model.model.attributes[index].params, key, true);
      }
    });
  });

  return {
    type: MODEL_FETCH_SUCCEEDED,
    model,
  };
}

export function postContentTypeSucceeded() {
  return {
    type: POST_CONTENT_TYPE_SUCCEEDED,
  };
}

export function resetShowButtonsProps() {
  return {
    type: RESET_SHOW_BUTTONS_PROPS,
  };
}

export function setButtonLoader() {
  return {
    type: SET_BUTTON_LOADER,
  };
}

export function submit() {
  return {
    type: SUBMIT,
  }
}

export function unsetButtonLoader() {
  return {
    type: UNSET_BUTTON_LOADER,
  };
}

export function updateContentType(data) {
  return {
    type: UPDATE_CONTENT_TYPE,
    data,
  };
}
