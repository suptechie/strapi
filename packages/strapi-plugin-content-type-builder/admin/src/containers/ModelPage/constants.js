/*
 *
 * ModelPage constants
 *
 */

export const DEFAULT_ACTION = 'ContentTypeBuilder/ModelPage/DEFAULT_ACTION';
export const MODEL_FETCH = 'ContentTypeBuilder/ModelPage/MODEL_FETCH';
export const MODEL_FETCH_SUCCEEDED = 'ContentTypeBuilder/ModelPage/MODEL_FETCH_SUCCEEDED';
