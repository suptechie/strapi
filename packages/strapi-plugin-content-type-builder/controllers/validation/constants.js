'use strict';

const CONTENT_TYPE = 'CONTENT_TYPE';
const COMPONENT = 'COMPONENT';

module.exports = {
  modelTypes: {
    CONTENT_TYPE,
    COMPONENT,
  },
};
