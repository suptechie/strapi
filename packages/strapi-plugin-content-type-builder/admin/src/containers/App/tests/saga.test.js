/**
 * Test  sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { all, fork, takeLatest, put } from 'redux-saga/effects';
import defaultSaga, { getData } from '../saga';

import { getDataSucceeded } from '../actions';
import { GET_DATA } from '../constants';

const response = {
  models: [
    { icon: 'fa-cube', name: 'permission', description: '', fields: 6, source: 'users-permissions' },
  ],
  allModels: [
    {
      collectionName: 'users-permissions_permission',
      connection: 'default',
      description: '',
      mainField: '',
      name: 'permission',
      attributes: [
        {
          name: 'type',
          params: { type: 'string', required: true, configurable: false },
        },
        {
          name: 'controller',
          params: { type: 'string', required: true, configurable: false },
        },
      ],
    },
  ],
};

describe('CTB <App /> GetData saga', () => {
  let getDataGenerator;

  beforeEach(() => {
    getDataGenerator = getData();
    const callDescriptor = getDataGenerator.next(response).value;

    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the getDataSucceeded action if it requests the data successfully', () => {
    const putDescriptor = getDataGenerator.next(response).value;
    
    expect(putDescriptor).toEqual(put(getDataSucceeded(response)));
  });

  it('should call the strapi.notification if the request fails', () => {
    const response = new Error('Some Error');
    getDataGenerator.throw(response).value;
  
    expect(strapi.notification.error).toHaveBeenCalled();
  });
});

describe('defaultSaga Saga', () => {
  const defaultSagaSaga = defaultSaga();

  it('should start a task to watch for GET_DATA', () => {
    const forkDescriptor = defaultSagaSaga.next().value;

    expect(forkDescriptor)
      .toEqual(
        all([
          fork(takeLatest, GET_DATA, getData),
        ]));
  }); 
});
