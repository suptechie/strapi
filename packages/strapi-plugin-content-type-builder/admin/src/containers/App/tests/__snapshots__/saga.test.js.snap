// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CTB <App /> GetData saga should call the strapi.notification if the request fails 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/content-type-builder/models",
      Object {
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`CTB <App /> GetData saga should dispatch the getDataSucceeded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/content-type-builder/models",
      Object {
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
